src/libraries/pretest.proj

```diff
@@ -68,11 +68,31 @@
              ContinueOnError="ErrorAndContinue" />
   </Target>
 
+  <Target Name="RunCrossgen2OnHost" Condition="'$(Crossgen2Host)' == 'true'">
+    <ItemGroup>
+      <FilesToCrossgen Include="$(TestHostRuntimePath)\Microsoft.*.dll" Exclude="$(TestHostRuntimePath)\Microsoft.CodeAnalysis*.dll;$(TestHostRuntimePath)\Microsoft.DiaSymReader.Native.*.dll" />
+      <FilesToCrossgen Include="$(TestHostRuntimePath)\System.*.dll" Exclude="$(TestHostRuntimePath)\System.Runtime.WindowsRuntime.dll" />
+    </ItemGroup>
+
+    <MakeDir Directories="$(TestHostRuntimePath)\crossgen2out" />
+    <Exec Command="$(TestHostRootPath)\dotnet $(TestHostRuntimePath)\crossgen2.dll %(FilesToCrossgen.FullPath) --targetarch=x64 -O --inputbubble -r:$(TestHostRuntimePath)\System.*.dll -r:$(TestHostRuntimePath)\Microsoft.*.dll -r:$(TestHostRuntimePath)\mscorlib.dll -r:$(TestHostRuntimePath)\netstandard.dll -o:$(TestHostRuntimePath)\crossgen2out\%(FilesToCrossgen.Filename)%(FilesToCrossgen.Extension)" 
+          IgnoreStandardErrorWarningFormat="true"/>
+  </Target>
+
+  <Target Name="MoveCrossgen2ResultsToHostRuntime" DependsOnTargets="RunCrossgen2OnHost">
+    <ItemGroup>
+      <FilesToCopy Include="$(TestHostRuntimePath)\crossgen2out\*.dll" />
+    </ItemGroup>
+
+    <Move SourceFiles="@(FilesToCopy)" DestinationFolder="$(TestHostRuntimePath)" />
+    <RemoveDir Directories="$(TestHostRuntimePath)\crossgen2out" />
```

Comment:
Can we use the same name that we used for System.Private.CoreLib which is/was `native` to store the crossgend assemblies. Also should we backup the IL version inside an `IL` folder and allow to get back to the IL state? Ie coverage measurement doesn't work with the crossgend assemblies and getting back to the non-crossgend state without rebuilding libraries would be desirable.

What about building crossgend without replacing the IL ones and allow to run on both modes with the switch? cc @ericstj

Reply:
We actually do the same thing in coreclr build-test.cmd - we just overwrite the assemblies in Core_Root with their crossgened versions and don't save anything. I think that keeping them in a different folder would be quite difficult as I think it would require to make a copy of the whole testhost folder and then changing all the references to the testhost in our scripts to this new one. 

We could possibly keep the IL versions, but it seems it would be tricky to manage that. At the beginning of test run without the crossgened assemblies, we would need to check if the IL folder was present and copy stuff from it back. That could result in stale files being copied - consider the following sequence:
* I build libraries with crossgening the testhost. That would create the IL folder with copies of the assemblies.
* I make some changes to coreclr or libraries and rebuild them without crossgening. Now the testhost contains updated non-crossgened assemblies. 
* I execute the tests - we find that there is an IL folder and copy the files back, overwriting the new ones with the stale ones. We cannot use time stamp of the files to check that, because we don't know if the files in the testhost are the crossgened ones or new IL ones.

I guess that we can find a way to make it work, but I doubt it is worth the hassle. I doubt that the flow when someone would need to run with and without crossgen without rebuilding libraries in between is important.

Reply:
OK, so what's the overall plan here? Will we add scheduled runs that set the property to true to test with crossgend?

Reply:
Right, that's what I would do once I fix the issues in crossgen2 that it has discovered.

Reply:
Btw, since the crossgen2out is just a temporary folder that's deleted after the crossgen completes, would you still want to rename it to native? I don't really care about the name.

Reply:
No if it's just temporary then it shouldn't matter. Thanks. If possible let's wait for ericstj's review.

Reply:
Sure, no need to rush.

Reply:
I think this is complex because you imagine modifying in place.  It’s actually easier if you imagine crossgen producing a new shared framework with a different root (eg: testhost_crossgen vs testhost).  Then it’s just a freshness check: is the non-CG newer? -> rerun crossgen.  In general we try make the build consume immutable assets and produce new immutable assets.  I’d prefer if we have a second folder.